### PR TITLE
Add `Vec::elementwise_min` and `Vec::elementwise_max` functions

### PR DESCRIPTION
**Pull request by fwyzard**
Return a vector where each element is set to the minimum (or maximum) of the elements in the same dimension of all the input vectors.

### CODE REVIEW

File: include/alpaka/vec/Vec.hpp [hunk 1]
```diff
@@ -402,6 +404,34 @@ namespace alpaka
             return r;
         }
 
+        //! \return The element-wise minimum of one or more vectors.
+        ALPAKA_NO_HOST_ACC_WARNING
+        template<typename... Vecs, typename = std::enable_if_t<boost::mp11::mp_same<Vec, Vecs...>::value>>
```
Comment (j-stephan) [inline include/alpaka/vec/Vec.hpp]: We have decided against the integration of Boost.Mp11 a while back. From what I can see you need it for `mp_same`. Shouldn't it be possible to use C++17 fold expressions here?

```suggestion
        template<typename... Vecs, typename = std::enable_if_t<std::is_same_v<Vec, Vecs> && ...>>
```

CC @bernhardmgruber

Comment (bernhardmgruber) [reply include/alpaka/vec/Vec.hpp]: Let me rephrase: We don't have consensus yet to allow the use of Boost.Mp11. The relevant discussion is here #1296.

In this particular instance, the fold expression suggested by Jan should work equally well!

File: include/alpaka/vec/Vec.hpp [hunk 2]
```diff
@@ -402,6 +404,34 @@ namespace alpaka
...
+                for(typename TDim::value_type i = 0; i < TDim::value; ++i)
+                    r[i] = std::min({p[i], qs[i]...});
+            }
+            return r;
+        }
+
+        //! \return The element-wise maximum of one or more vectors.
+        ALPAKA_NO_HOST_ACC_WARNING
+        template<typename... Vecs, typename = std::enable_if_t<boost::mp11::mp_same<Vec, Vecs...>::value>>
+        ALPAKA_FN_HOST_ACC friend constexpr auto elementwise_max(Vec const& p, Vecs const&... qs) -> Vec
```
Comment (bernhardmgruber) [inline include/alpaka/vec/Vec.hpp]: Since you seem to intend to call this functions via a qualified name `alpaka::elementwise_max` and not ADL-only: please move it out of the definition of the `Vec` class and make it an ordinary free function. This should also make your unit test compile.

**Review by bernhardmgruber — CHANGES_REQUESTED** [review]

File: include/alpaka/vec/Vec.hpp (hunk 2 above)
Comment (fwyzard) [reply include/alpaka/vec/Vec.hpp]: Ah... I read the first part of

> A name first declared in a friend declaration within a class or class template `X` becomes a member of the innermost enclosing namespace of `X`, ...

but  I missed the second part

> ... but is not visible for lookup (except argument-dependent lookup that considers `X`) unless a matching declaration at namespace scope is provided.

File: include/alpaka/vec/Vec.hpp (hunk 1 above)
Comment (fwyzard) [reply include/alpaka/vec/Vec.hpp]: OK, I see.

Thanks, the fold expression does seem to work well (it only needs a couple of `(...)`).

File: include/alpaka/vec/Vec.hpp (hunk 2 above)
Comment (bernhardmgruber) [reply include/alpaka/vec/Vec.hpp]: Ok, now you added declarations of the functions outside `Vec`. Why not move the entire definition outside as well? That would reduce the amount of code.

File: include/alpaka/vec/Vec.hpp [hunk 3]
```diff
@@ -402,6 +402,34 @@ namespace alpaka
             return r;
         }
 
+        //! \return The element-wise minimum of one or more vectors.
+        ALPAKA_NO_HOST_ACC_WARNING
+        template<typename... Vecs, typename = std::enable_if_t<(std::is_same_v<Vec, Vecs> && ...)>>
+        ALPAKA_FN_HOST_ACC friend constexpr auto elementwise_min(Vec const& p, Vecs const&... qs) -> Vec
```
Comment (bernhardmgruber) [inline include/alpaka/vec/Vec.hpp]:
```suggestion
        template<typename... Vecs, typename = std::enable_if_t<(std::is_same_v<Vec, Vecs> && ... && true)>>
        ALPAKA_FN_HOST_ACC friend constexpr auto elementwise_min(Vec const& p, Vecs const&... qs) -> Vec
```
To support `elementwise_min` with a single argument.

File: test/unit/vec/src/VecTest.cpp [hunk 4]
```diff
@@ -262,6 +262,32 @@ TEST_CASE("basicVecTraits", "[vec]")
...
+            constexpr auto vecMin = alpaka::elementwise_min(vec, vec3);
+            STATIC_REQUIRE(std::is_same_v<alpaka::Dim<std::decay_t<decltype(vecMin)>>, Dim>);
+            STATIC_REQUIRE(std::is_same_v<alpaka::Idx<std::decay_t<decltype(vecMin)>>, bool>);
+
+            constexpr alpaka::Vec<Dim, Idx> referenceVec(
+                static_cast<Idx>(0u),
+                static_cast<Idx>(8u),
+                static_cast<Idx>(3u));
+            STATIC_REQUIRE(referenceVec == vecMin);
+        }
```
Comment (bernhardmgruber) [inline test/unit/vec/src/VecTest.cpp]: Consider adding tests for calling `elementwise_min` with a single argument.
Please also add a test calling it with more than 2 arguments. Thx!

File: include/alpaka/vec/Vec.hpp (hunk 2 above)
Comment (fwyzard) [reply include/alpaka/vec/Vec.hpp]: I assumed they still need to be declared as `friend` to access the `Vec`'s members ?

Comment (bernhardmgruber) [reply include/alpaka/vec/Vec.hpp]: Yes they would, but IIUC you are not accessing anything private.

**Comment by bernhardmgruber** [comment]
Please also rebase your PR on the latest `develop`!

File: include/alpaka/vec/Vec.hpp (hunk 3 above)
Comment (fwyzard) [reply include/alpaka/vec/Vec.hpp]: Actually that's not needed:

> When a unary fold is used with a pack expansion of length zero, only the following operators are allowed:
>  1) **Logical AND (`&&`)**. The value for the empty pack is `true`
>  2) Logical OR (`||`). The value for the empty pack is `false`
>  3) The comma operator (`,`). The value for the empty pack is `void()`


But it's good to check - I'm adding a test with a single argument to the unit test.

File: test/unit/vec/src/VecTest.cpp (hunk 4 above)
Comment (fwyzard) [reply test/unit/vec/src/VecTest.cpp]: Done.

File: include/alpaka/vec/Vec.hpp (hunk 2 above)
Comment (fwyzard) [reply include/alpaka/vec/Vec.hpp]: 👍🏻

**Review by bernhardmgruber — APPROVED** [review]
